### PR TITLE
JDBC connection cannot be aquired

### PR DESCRIPTION
**Pull request by manandearth**
>clojure.lang.ExceptionInfo: java.sql.SQLException in Interceptor  - Connections could not be acquired from the underlying database!
,,,
>Caused by: com.mchange.v2.resourcepool.CannotAcquireResourceException: A ResourcePool could not acquire a resource from its primary factory or source.
,,, 
>Caused by: org.postgresql.util.PSQLException: This ResultSet is closed.

### CODE REVIEW

File: dev/dev.clj [hunk 1]
```diff
@@ -17,18 +17,18 @@
...
 (defn dev-system
   []
   (component/system-map
    :service-map playground.server/dev-map
-   :background-processor (background-processor/new :queue-name "cljtest")
-   :enqueuer (enqueuer/new :queue-name "cljtest")
-   :db (modular.postgres/map->Postgres {:url "jdbc:postgresql:ebdb" :user "vemv" :password ""})
+   ;; :background-processor (background-processor/new :queue-name "cljtest")
+   ;; :enqueuer (enqueuer/new :queue-name "cljtest")
+   :db (modular.postgres/map->Postgres {:url "jdbc:postgresql:swallows" :user "swallows" :password "swallows"})
```
Comment (manandearth) [inline dev/dev.clj:31]: Doesn't work with this :db map (tried long address same result) 
looked into the modular.postgres ns. I don't understand how pooling work. (related?)

File: src/playground/service.clj [hunk 2]
```diff
@@ -71,7 +75,9 @@
     ["/api" :get (into component-interceptors [http/json-body (param-spec-interceptor ::api :query-params) `api])]
     ["/invoices/insert" :get (into component-interceptors [http/json-body (param-spec-interceptor ::invoices.insert/api :query-params) `invoices.insert/perform])]
     ["/invoices/:id" :get (into component-interceptors [http/json-body (param-spec-interceptor ::invoices.retrieve/api :path-params) `invoices.retrieve/perform])]
```
Comment (manandearth) [inline src/playground/service.clj:77]: /invoices/:id is the route I'd like to test. I created  `CREATE TABLE users (id INTEGER, email VARCHAR(50));` in the DB.

File: src/playground/service.clj [hunk 3]
```diff
@@ -71,7 +75,9 @@
...
-    ["/invoices/delete" :get (into component-interceptors [http/json-body `invoices.delete/perform])]})
+    ["/invoices/delete" :get (into component-interceptors [http/json-body `invoices.delete/perform])]
+    ;; ["/nests/:species" :get (into component-interceptors [http/json-body (param-spec-interceptor ::invoices.retrieve/api :path-params) `nests.retrieve/perform])]
```
Comment (manandearth) [inline src/playground/service.clj:79]: similar route to "/invoices/:id" to query the `nests` table WIP

File: src/playground/services/nests/retrieve/endpoint.clj [hunk 4]
```diff
@@ -0,0 +1,27 @@
+(ns playground.services.nests.retrieve.endpoint
```
Comment (manandearth) [inline src/playground/services/nests/retrieve/endpoint.clj:1]: copied and updated endpoint for `:nests` table . WIP (Spec in progress..)

File: project.clj [hunk 5]
```diff
@@ -28,7 +29,10 @@
   :repl-options {:port 41234}
   :min-lein-version "2.0.0"
   :resource-paths ["config" "resources"]
-  :profiles {:dev {:dependencies [[io.pedestal/pedestal.service-tools "0.5.3"]]
+  :profiles {:dev {:dependencies [[io.pedestal/pedestal.service-tools "0.5.3"]
+                                  [com.stuartsierra/component.repl "0.2.0"]
+                                  [org.clojure/tools.namespace "0.2.11"]
```
Comment (vemv) [inline project.clj:34]: Use tools.namespace 0.3.0-alpha4, it's been the 'good one' for a few years despite the alpha status

(it's a stupid intricacy, you're not at fault)

File: src/playground/service.clj (hunk 3 above)
Comment (vemv) [reply src/playground/service.clj:79]: Don't add WIPs though

Remember: tackle one problem at a time, minimizing noise for the code reviewer

**Comment by vemv** [comment]
Funnily enough I can reproduce the error, even after locally restoring my credentials:

image

but it's our machines which are buggy, not the code itself.

If I visit `http://localhost:8080/invoices/4`, we'll get a damn long stacktrace, with the real reason buried somewhere:

```
Caused by: org.postgresql.util.PSQLException: FATAL: role "vemv" does not exist
```

In my case, since creating `playground` I swapped laptops, and no `vemv` role (user) exists in my psql.

Very likely you have the same issue. Check the stacktrace, and ensure that ` {:url "jdbc:postgresql:swallows" :user "swallows" :password "swallows"})` reflects a config that does exist.

**Comment by manandearth** [comment]
I get something different. I believe that my credentials are correct since when I change password it gives a `wrong password` message.

> clojure.lang.ExceptionInfo:

> java.sql.SQLException in Interceptor - Connections could not be acquired from the underlying database!
,,,
>Caused by: com.mchange.v2.resourcepool.CannotAcquireResourceException: A ResourcePool could not acquire a resource from its primary factory or source.
,,,

>Caused by: com.mchange.v2.resourcepool.CannotAcquireResourceException: A ResourcePool could not acquire a resource from its primary factory or source.

,,,

>Caused by: org.postgresql.util.PSQLException: This ResultSet is closed.
-------------------------------------

**Comment by vemv** [comment]
So, you have a `users` table right?

Can you try inserting at least one row from the `psql` cli?

**Comment by manandearth** [comment]
Yes from psql and I can probably also from jdbc with a full connection-uri string.
It's either the pool library or the `spec-interceptor`, I don't know how to address that.

\conninfo gives:
`You are connected to database "swallows" as user "swallows" via socket in "/run/postgresql" at port "5432".`

**Comment by vemv** [comment]
http://localhost:8080/invoices/4 it's working for me after fixing the mentioned username.

Please paste here:

* the full stacktrace, as a gist link (since it's too long)
* the result of `select * from users` in a `psql swallows -U swallows` session (note that the session connects to the swallows db, and under the swallows username), as a code block (three backticks)

**Comment by manandearth** [comment]
[stacktrace gist link](https://gist.github.com/manandearth/c49fcd97dca50936bb3c455408f25215)

psql input when `\conninfo` is:
```
You are connected to database "swallows" as user "swallows" via socket in "/run/postgresql" at port "5432".
```
is:
```
swallows=> select * from users;                                        
 id |     email                                                        
----+---------------
  3 | this@that.com
  2 | is@what.com
(2 rows)
```

**Comment by vemv** [comment]
Great answer, ty!

What's your output for this?

```
$ psql --version
psql (PostgreSQL) 9.6.10
```

**Comment by manandearth** [comment]
psql (PostgreSQL) 10.4

**Comment by vemv** [comment]
yeah, suspected so. I recall the modular psql component is dated by now, with no newer version available.

You'll have to downgrade to 9.x, preferably 9.6

Sorry for the nuisance!

**Comment by manandearth** [comment]
might be easier to update the modular library if we knew what needs updating looks like a straight forward jdbc interop.

**Comment by vemv** [comment]
There's no such update; Juxt's `modular` project has stalled

There's probably a nice solution but it'd certainly take some research+work.

**Pull request merged**